frb_codegen/src/ir/file.rs

```diff
@@ -1,29 +1,123 @@
+use convert_case::{Case, Casing};
+
 use crate::target::Target;
-use crate::utils::misc::mod_from_rust_path;
+use crate::utils::misc::{is_multi_blocks_case, mod_from_rust_path, BlockIndex};
 use crate::{generator, ir::*, Opts};
 use std::collections::{HashMap, HashSet};
+use std::thread;
+
+use crate::utils::misc::ExtraTraitForVec;
 
 pub type IrStructPool = HashMap<String, IrStruct>;
 pub type IrEnumPool = HashMap<String, IrEnum>;
 
+use std::cell::RefCell;
+
+// this variable should be `none` for single block case, and not none for multi-blocks case.
+thread_local!(pub static SHARED_MODULE: RefCell<Option<String>> = RefCell::new(None));
+
+// shared types among all regular API blocks, in context of function paramters
+thread_local!(pub static SHARED_TYPES_INPUT: RefCell<HashSet<IrType>> = RefCell::new(HashSet::new()));
+thread_local!(static FETCHED_FOR_SHARED_TYPES_INPUT: RefCell<bool> = RefCell::new(false));
+
+// shared types among all regular API blocks in context of function outputs
+thread_local!(static SHARED_TYPES_OUTPUT: RefCell<HashSet<IrType>> = RefCell::new(HashSet::new()));
+thread_local!(static FETCHED_FOR_SHARED_TYPES_OUTPUT: RefCell<bool> = RefCell::new(false));
+
+// shared types among all regular API blocks
+thread_local!(static SHARED_TYPES_ALL: RefCell<HashSet<IrType>> = RefCell::new(HashSet::new()));
+thread_local!(static FETCHED_FOR_SHARED_TYPES_ALL: RefCell<bool> = RefCell::new(false));
```

Comment:
oops so many global variables... it would be great if we can avoid them! (methods are discussed in another comment above)